pkg/vf/virtio.go

```diff
@@ -289,6 +292,112 @@ func (dev *VirtioVsock) AddToVirtualMachineConfig(vmConfig *VirtualMachineConfig
 	return nil
 }
 
+func (dev *NetworkBlockDevice) toVz() (vz.StorageDeviceConfiguration, error) {
+	if err := dev.validateNbdURI(dev.URI); err != nil {
+		return nil, fmt.Errorf("invalid NBD device 'uri': %s", err.Error())
+	}
+
+	if err := dev.validateNbdDeviceIdentifier(dev.DeviceIdentifier); err != nil {
+		return nil, fmt.Errorf("invalid NBD device 'deviceId': %s", err.Error())
+	}
+
+	attachment, err := vz.NewNetworkBlockDeviceStorageDeviceAttachment(dev.URI, dev.Timeout, dev.ReadOnly, dev.SyncronizationModeVZ())
+	if err != nil {
+		return nil, err
+	}
+
+	vzdev, err := vz.NewVirtioBlockDeviceConfiguration(attachment)
+	if err != nil {
+		return nil, err
+	}
+	err = vzdev.SetBlockDeviceIdentifier(dev.DeviceIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	return vzdev, nil
+}
+
+func (dev *NetworkBlockDevice) validateNbdURI(uri string) error {
+	if uri == "" {
+		return fmt.Errorf("'uri' must be specified")
+	}
+
+	parsed, err := url.Parse(uri)
+	if err != nil {
+		return fmt.Errorf("error: %w", err)
+	}
+
+	// The format specified by https://github.com/NetworkBlockDevice/nbd/blob/master/doc/uri.md
+	if parsed.Scheme != "nbd" && parsed.Scheme != "nbds" && parsed.Scheme != "nbd+unix" && parsed.Scheme != "nbds+unix" {
+		return fmt.Errorf("invalid scheme: %s. Expected one of: 'nbd', 'nbds', 'nbd+unix', or 'nbds+unix'", parsed.Scheme)
+	}
+
+	return nil
+}
+
+func (dev *NetworkBlockDevice) validateNbdDeviceIdentifier(deviceID string) error {
+	if deviceID == "" {
+		return fmt.Errorf("'deviceId' must be specified")
+	}
+
+	if strings.Contains(deviceID, "/") {
+		return fmt.Errorf("invalid 'deviceId': it cannot contain any forward slash")
+	}
+
+	if len(deviceID) > 255 {
+		return fmt.Errorf("invalid 'deviceId': exceeds maximum length")
+	}
+
+	return nil
+}
+
+func (dev *NetworkBlockDevice) SyncronizationModeVZ() vz.DiskSynchronizationMode {
+	if dev.SynchronizationMode == config.SynchronizationNoneMode {
+		return vz.DiskSynchronizationModeNone
+	}
+	return vz.DiskSynchronizationModeFull
+}
+
+func (dev *NetworkBlockDevice) AddToVirtualMachineConfig(vmConfig *VirtualMachineConfiguration) error {
+	storageDeviceConfig, err := dev.toVz()
+	if err != nil {
+		return err
+	}
+	log.Infof("Adding NBD device (uri: %s, deviceId: %s)", dev.URI, dev.DeviceIdentifier)
+	vmConfig.storageDevicesConfiguration = append(vmConfig.storageDevicesConfiguration, storageDeviceConfig)
+
+	return nil
+}
+
+func ListenNetworkBlockDevices(vm *VirtualMachine) error {
+	for _, dev := range vm.vfConfig.storageDevicesConfiguration {
+		attachment := dev.Attachment()
+		if nbdAttachment, isNbdAttachment := attachment.(*vz.NetworkBlockDeviceStorageDeviceAttachment); isNbdAttachment {
+			nbdID, err := dev.(*vz.VirtioBlockDeviceConfiguration).BlockDeviceIdentifier()
+			if err != nil {
+				log.Info("Found NBD device with no identifier. Please set a deviceId to each NBD device.")
+				return fmt.Errorf("NetworkBlockDevice must have deviceId. Error %v", err)
```

Comment:
you could use `%w` instead of `%v` so that the error is wrapped in the returned error object instead of only `err.String()` being added to the error string.